File: go/solutions/day5.go
package solutions

import (
	"strconv"
	"strings"

	"main.go/go/helpers"
)

func Day5() (string, string) {
	diagram, instructions := parseInput()

	// Part 1
	stacks := fillStacks(diagram)
	rearrange(stacks, instructions, false)
	part1 := topOfEachStack(stacks)

	// Part 2
	stacks = fillStacks(diagram)
	rearrange(stacks, instructions, true)
	part2 := topOfEachStack(stacks)
	return part1, part2
}

func parseInput() ([]string, []string) {
	file, scanner := helpers.GetFile(5)
	defer file.Close()

	diagram := []string{}
	instructions := []string{}
	diagramComplete := false
	for scanner.Scan() {
		line := scanner.Text()
		if diagramComplete {
			instructions = append(instructions, line)
		} else if line == "" {
			diagramComplete = true
		} else {
			diagram = append([]string{line}, diagram...)
		}
	}
	return diagram, instructions
}

func fillStacks(diagram []string) map[string][]string {
	stacks := make(map[string][]string)
	for k, line := range diagram {
		i, j := 1, 1
		for i < len(line) {
			crate := string(line[i])
			if k == 0 {
				stacks[crate] = []string{}
			} else if crate != " " {
				key := strconv.Itoa(j)
				stacks[key] = append(stacks[key], crate)
			}
			i += 4
			j++
		}
	}
	return stacks
}

func rearrange(stacks map[string][]string, instructions []string, inBulk bool) {
	for _, instruction := range instructions {
		parts := strings.Fields(instruction)
		quantity, _ := strconv.Atoi(parts[1])
		from := parts[3]
		to := parts[5]
		to_stack := stacks[to]
		from_stack := stacks[from]
		if inBulk {
			index := len(from_stack) - quantity
			to_stack = append(to_stack, from_stack[index:]...)
			from_stack = from_stack[:index]
		} else {
			for i := 0; i < quantity; i++ {
				index := len(from_stack) - 1
				to_stack = append(to_stack, from_stack[index:]...)
				if len(from_stack) > 1 {
					from_stack = from_stack[:index]
				} else {
					from_stack = []string{}
				}
			}
		}
		stacks[from], stacks[to] = from_stack, to_stack
	}
}

func topOfEachStack(stacks map[string][]string) string {
	tops := []string{}
	for i := 1; i <= len(stacks); i++ {
		stack := stacks[strconv.Itoa(i)]
		if len(stack) > 0 {
			tops = append(tops, stack[len(stack)-1])
		}
	}
	return strings.Join(tops, "")
}

File: python/day5.py
from utils import get_data


def main():
    data = get_data(5)
    crate_mover_9000, crate_mover_9001 = CrateMover9000(
        data), CrateMover9001(data)
    crate_mover_9000.rearrange()
    crate_mover_9001.rearrange()
    return crate_mover_9000.top_of_each_stack, crate_mover_9001.top_of_each_stack


class CrateMover9000:
    def __init__(self, data):
        sep = data.index('')
        self.diagram, self.instructions = data[:sep], data[sep + 1:]
        stacks = {}
        for k, line in enumerate(reversed(self.diagram)):
            i = j = 1
            while i < len(line):
                crate = line[i].strip()
                if k == 0:
                    stacks[crate] = []
                elif crate:
                    stacks[str(j)].append(crate)
                i += 4
                j += 1
        self.stacks = stacks

    def rearrange(self):
        for instruction in self.instructions:
            _, quantity, _, from_stack, _, to_stack = instruction.split(' ')
            self.perform_move(int(quantity), from_stack, to_stack)

    def perform_move(self, quantity, from_stack, to_stack):
        for _ in range(quantity):
            self.stacks[to_stack].append(self.stacks[from_stack].pop())

    @property
    def top_of_each_stack(self):
        return ''.join([s[-1] for s in self.stacks.values() if len(s)])


class CrateMover9001(CrateMover9000):
    def perform_move(self, quantity, from_stack, to_stack):
        index = len(self.stacks[from_stack]) - quantity
        self.stacks[to_stack] += self.stacks[from_stack][index:]
        self.stacks[from_stack] = self.stacks[from_stack][:index]


if __name__ == '__main__':
    print(main())
